cdap-app-fabric/src/main/java/co/cask/cdap/app/mapreduce/MapReduceMetricsInfo.java

```
@@ -0,0 +1,217 @@
+                getAggregates(reduceContext, MapReduceMetrics.METRIC_OUTPUT_RECORDS));
+
+    float mapProgress = getAggregates(mapContext, MapReduceMetrics.METRIC_COMPLETION) / 100.0F;
+    float reduceProgress = getAggregates(reduceContext, MapReduceMetrics.METRIC_COMPLETION) / 100.0F;
+
+    return new MRJobInfo(null, null, null, mapProgress, reduceProgress, metrics, mapTaskInfos, reduceTaskInfos);
+  }
+
+  private String prependSystem(String metric) {
+    return "system." + metric;
+  }
+
+  private long getAggregates(Map<String, String> context, String metric) throws Exception {
+    MetricDataQuery metricDataQuery =
+      new MetricDataQuery(0, Integer.MAX_VALUE, Integer.MAX_VALUE, prependSystem(metric), MetricType.COUNTER, context,
+                          ImmutableList.<String>of());
+    Collection<MetricTimeSeries> query = metricStore.query(metricDataQuery);
+    if (query.isEmpty()) {
+      return 0;
+    }
+    MetricTimeSeries timeSeries = Iterables.getOnlyElement(query);
+    List<TimeValue> timeValues = timeSeries.getTimeValues();
+    TimeValue timeValue = Iterables.getOnlyElement(timeValues);
+    return timeValue.getValue();
+  }
+
+  // queries MetricStore for one metric across all tasks of a certain TaskType, using GroupBy InstanceId
+  private Map<String, Long> queryGroupedAggregates(Map<String, String> context, String metric) throws Exception {
+    MetricDataQuery metricDataQuery =
+      new MetricDataQuery(0, Integer.MAX_VALUE, Integer.MAX_VALUE, prependSystem(metric), MetricType.GAUGE, context,
+                          ImmutableList.of(Constants.Metrics.Tag.INSTANCE_ID));
+    Collection<MetricTimeSeries> query = metricStore.query(metricDataQuery);
+
+    // runId -> metricValue
+    Map<String, Long> taskMetrics = Maps.newHashMap();
+    for (MetricTimeSeries metricTimeSeries : query) {
+      List<TimeValue> timeValues = metricTimeSeries.getTimeValues();
+      TimeValue timeValue = Iterables.getOnlyElement(timeValues);
```

do you want to throw the exception if no element is present or (log and skip) ?